drivers/rtc/rtc_am1805.c

```diff
@@ -0,0 +1,601 @@
+/* To get from the alarm registers */
+static int am1805_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+		struct rtc_time *timeptr)
+{
+
+	int err;
+	uint8_t regs[6];
+
+	struct am1805_data *data = dev->data;
+	const struct am1805_config *config = dev->config;
+
+	if (id != 0U) {
+		LOG_ERR("invalid ID %d", id);
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	err = i2c_burst_read_dt(&config->int_i2c, REG_ALM_SECONDS_ADDR, regs, sizeof(regs));
+	if (err != 0) {
+		goto unlock;
+	}
+
+	timeptr->tm_sec = bcd2bin(regs[0] & SECONDS_BITS);
+	timeptr->tm_min = bcd2bin(regs[1] & MINUTES_BITS);
+	timeptr->tm_hour = bcd2bin(regs[2] & HOURS_BITS);
+	timeptr->tm_mday = bcd2bin(regs[3] & DATE_BITS);
+	timeptr->tm_mon = bcd2bin(regs[4] & MONTHS_BITS);
+	timeptr->tm_wday = bcd2bin(regs[5] & WEEKDAY_BITS);
+
+	*mask = (AM1805_RTC_ALARM_TIME_MASK);
+
+	LOG_DBG("get alarm: wday = %d, mon = %d, mday = %d, hour = %d, min = %d, sec = %d, "
+			"mask = 0x%04x",
+			timeptr->tm_wday, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_hour,
+			timeptr->tm_min, timeptr->tm_sec, *mask);
+
+unlock:
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int am1805_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+		const struct rtc_time *timeptr)
+{
+	int err;
+	uint8_t regs[6];
+
+	struct am1805_data *data = dev->data;
+	const struct am1805_config *config = dev->config;
+
+	if (id != 0U) {
+		LOG_ERR("invalid ID %d", id);
+		return -EINVAL;
+	}
+
+	if ((mask & ~(AM1805_RTC_ALARM_TIME_MASK)) != 0U) {
+		LOG_ERR("unsupported alarm field mask 0x%04x", mask);
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	/* Disable timer control registers before the initialization */
+	err = i2c_reg_update_byte_dt(&config->int_i2c, REG_TIMER_CTRL_ADDR,
+					REG_TIMER_CTRL_RPT_BITS, 0);
+	if (err != 0) {
+		goto unlock;
+	}
+
+	/* Clear the interrupt mask for alarm */
+	err = i2c_reg_update_byte_dt(&config->int_i2c, REG_IRQ_MASK_ADDR,
+					REG_IRQ_MASK_AIE, 0);
+	if (err != 0) {
+		goto unlock;
+	}
+
+	/* Clear the status bit */
+	err = i2c_reg_update_byte_dt(&config->int_i2c, REG_STATUS_ADDR,
+					REG_STATUS_ALM, 0);
+	if (err != 0) {
+		goto unlock;
+	}
+
+	/* When mask is 0 */
+	if (mask == 0) {
+		LOG_DBG("The alarm is disabled");
+		goto unlock;
+	}
+
+	regs[0] = bin2bcd(timeptr->tm_sec) & SECONDS_BITS;
+	regs[1] = bin2bcd(timeptr->tm_min) & MINUTES_BITS;
+	regs[2] = bin2bcd(timeptr->tm_hour) & HOURS_BITS;
+	regs[3] = bin2bcd(timeptr->tm_mday) & DATE_BITS;
+	regs[4] = bin2bcd(timeptr->tm_mon) & MONTHS_BITS;
+	regs[5] = bin2bcd(timeptr->tm_wday) & WEEKDAY_BITS;
+
+	LOG_DBG("set alarm: second = %d, min = %d, hour = %d, mday = %d, month = %d,"
+			"wday = %d,  mask = 0x%04x",
+			timeptr->tm_sec, timeptr->tm_min, timeptr->tm_hour, timeptr->tm_mday,
+			timeptr->tm_mon, timeptr->tm_wday, mask);
+
+	err = i2c_burst_write_dt(&config->int_i2c, REG_ALM_SECONDS_ADDR, regs, sizeof(regs));
```

Comment:
@fabiobaltieri , I already came across that in the documentation, but isn't that the responsibility of the i2c controller or the subsystem ? IMO,Thought of just to stick with the i2c api exposed to the driver development.

Reply:
I mean, that is the documentation of the i2c subsystem telling you that the controller may not able to handle it. Not sure what you mean by "stick with the i2c api exposed to the driver development". That is the API, those are the limitations.

Anyway the only case I bumped into was for a large write op with the nordic twim driver, and I think that even that could have been fixed with some tweak. Fine by me if you want to leave this, just pointing out it may have to be reworked down the road.

Reply:
>Fine by me if you want to leave this, just pointing out it may have to be reworked down the road. 

@fabiobaltieri , Sure.